engine/crates/engine-v2/schema/src/introspection.rs

```diff
@@ -0,0 +1,416 @@
+use std::{
+    collections::HashMap,
+    ops::{Deref, DerefMut},
+};
+
+use crate::{
+    Definition, EnumId, EnumValue, Field, FieldArgument, FieldId, FieldType, FieldTypeId, ObjectField, ObjectId,
+    ScalarId, Schema, StringId, Value, Wrapping,
+};
+
+pub struct IntrospectionFields<'a> {
+    schema: &'a mut Schema,
+    strings_map: HashMap<String, StringId>,
+}
+
+impl<'a> Deref for IntrospectionFields<'a> {
+    type Target = Schema;
+    fn deref(&self) -> &Self::Target {
+        self.schema
+    }
+}
+
+impl<'a> DerefMut for IntrospectionFields<'a> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.schema
+    }
+}
+
+impl<'a> IntrospectionFields<'a> {
+    // /!\ You MUST call schema.ensure_proper_ordering() after calling this function. /!\
```

Comment:
Seems like you could enforce this by taking ownership of `Schema` and having a `.finish()` function that calls `ensure_proper_ordering()` then returns ownership of the schema.

Probably safer than assuming people will read this comment

Reply:
hum indeed, good point

Reply:
fixed in next pr